doc: add Resumable Full Refresh documentation. (#41334)
airbytehq/airbyte: docs/operator-guides/resumability.md

Although incremental syncs are superior in all cases, they require implementing a cursor, which is not always possible.

Consider a data analyst without direct access to the underlying database. Not being able to introspect schemas makes identifying and selecting the correct cursor column a challenge.
In fact, there might not be a cursor column to start! Full Refresh, in this case, is the only option.

### Artificial Cursors

Enabling Resumable Full Refresh requires **Artificial Cursors**. An Artificial Cursor is a best-effort attempt to construct a relatively stable cursor.
These cursors are artificial as they are often not present in the data schema, e.g. Postgres's CTID is a system-level column, and subject to independent
change i.e. without external control or input. This differs from a proper cursor column, e.g. an `update_at` column, that is present in the data schema,
and changes in response to user input.

### Enabling Resumable Full Refresh

Resuamble Full Refresh is automatically enabled on connections with:
1) A Resumable-Full-Refresh-enabled Source
2) A Refresh-enabled Destination

The following Database Sources support Resumable Full Refresh as of 9th July 2024:

> 1) **Source MySql >= 3.4.7**

Author: @xiaohansong are these versions correct?

---
products: all
---

# Resumability & Resumable Full Refresh

:::info

> This page describes a coming-soon feature. Resumable Full Refresh will be available as of Airbyte v1.0.0.

Reviewer: https://docs.airbyte.com/contributing-to-airbyte/#documentation States we shouldn't pre-announce. Do we have a timeline for the availability of this feature?

Reviewer: Maybe, instead, you can link to the section that lists what sources and APIs support resumable full refresh with something a little like "This feature is currently limited to certain sources and APIs. The section on [enabling resumable full refresh](link it) lists those for which this feature is currently supported"

Author: RFR is currently available in Cloud for the supported connectors. It will be available in OSS in 2 - 3 weeks.

I like the suggestion to link to a section where the feature is supported. Is it ok if the linked section is part of this document?

Author: I didn't complete my thought: Happy to leave this in a PR!

Thought it would be useful to have this out since the feature is available in various product lines.

:::

## Resumability

Airbyte strives to create resilient syncs that can handle many types of issues (for example, networking issues, flaky source APIs or under provisioned
destination servers). Whenever possible, Airbyte syncs are resumable. “Resumable” in this context means that if something goes wrong in the first attempt
of your sync, we will immediately try again with a subsequent attempt - in fact, Airbyte will keep retrying as long as the source is producing new records
and the destination can commit them. In order to achieve this resumability, Airbyte may deliver a record more than once - **Airbyte is a deliver at-least-once system**.

Consider a Postgres source. We could build a query that does `select * from users`, which would guarantee that each row is emitted only once, but that would lead to some problems:

> - First, it would keep a long-lasting transaction lock on the users table for the whole duration of the sync which would slow down the source

Reviewer: ```suggestion
- First, it would keep a long-lasting transaction lock on the `users` table for the duration of the sync.  This would slow down the source.
```

> - Tables over a certain size won’t be able to fit in memory, again dramatically slowing down your database while syncing (or even crashing the database)

Reviewer: ```suggestion
- Tables over a certain size would not be able to fit in memory. This would dramatically slow down the database while syncing (or even crashing) the database.
```

> - There’s no logical cursor that we can resume from if we need to retry the sync

Reviewer: ```suggestion
- There is no logical cursor that Airbyte can use as a placeholder to resume from if we need to retry the sync.
```

> So, to get around all of these issues, we instead issue queries that look more like this:

Reviewer: ```suggestion
To get around all of these issues, we issue queries that look like this:
```

- `select * from users where CTID >= X AND CTID < Y`
- `select * from users where CTID >= Y AND CTID < Z`
- `...`

> CTID is a special postgres column that we use to paginate the contents of the table, and this affords us a cursor we can use to break the query into chunks, both keeping the database happy, and allowing Airbyte to resume the sync half-way though if needed. However, if the sync only got partially through a query, the records in that query will be re-sent on the second attempt. If you want your destination to only contain a unique instance for each record, please choose a destination and sync mode that includes deduplication.

Reviewer: ```suggestion
CTID is a special Postgres column that we use to paginate the contents of a table. This provides us with a cursor, allowing us to break the query into chunks. Doing this keeps the database happy, and allows Airbyte to resume the sync from the point indicated by the cursor if needed. Note that if the sync gets partially through a query, the records in that query would be re-sent on the second attempt, resulting in duplicate records. If you want your destination to only contain a unique instance for each record, you should choose a destination and sync mode that includes deduplication.
```

This example uses a database source, but the same logic holds for API sources as well. A series of requests to a user's API endpoint might be:

- `curl -x GET api.com/v1/users?page=1`
- `curl -x GET api.com/v1/users?page=2`
- `...`

> Resumability applies to all sync modes, including standard Full Refresh + Overwrite and Full Refresh + Append syncs, along with [Refreshes](./refreshes.md).

Author: thank you! I couldn't figure this out so I left it for later.

## Resumable Full Refresh

> Resumability is an inherent feature of Incremental syncs. What does it mean for Full Refresh syncs?

Reviewer: ```suggestion
Resumability is an inherent feature of Incremental syncs. 
```

> ### Why Full Refresh?

Reviewer: ```suggestion

```

> Although incremental syncs are superior in all cases, they require a cursor, which is not always possible.

Reviewer: ```suggestion
Although incremental syncs are superior in all cases, they require implementing a cursor, which is not always possible.
```

Consider a data analyst without direct access to the underlying database. Not being able to introspect schemas makes identifying and selecting the correct cursor column a challenge.

> In fact, there might not be a cursor column to start! Full Refresh is the only option.

Reviewer: ```suggestion
In fact, there might not be a cursor column to start! Full Refresh, in this case, is the only option.
```

### Artificial Cursors

Enabling Resumable Full Refresh requires **Artificial Cursors**. An Artificial Cursor is a best-effort attempt to construct a relatively stable cursor.
These cursors are artificial as they are often not present in the data schema, e.g. Postgres's CTID is a system-level column, and subject to independent

> change i.e. without external control or input. Contrast this with a proper cursor column, e.g. an `update_at` column, that is present in the data schema,

Reviewer: ```suggestion
change i.e. without external control or input. This differs from a proper cursor column, e.g. an `update_at` column, that is present in the data schema,
```
